presto-main-base/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java

```diff
@@ -1437,11 +1440,66 @@ else if (character == '%') {
     @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND)
     public static long parseDuration(@SqlType("varchar(x)") Slice duration)
     {
+        String durationStr = duration.toStringUtf8().trim();
+
+        if (durationStr.isEmpty()) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "duration is empty");
+        }
+
         try {
-            return Duration.valueOf(duration.toStringUtf8()).toMillis();
+            Pattern pattern = Pattern.compile("^(\\d+(?:\\.\\d+)?)\\s*([a-zA-Z]+)$");
+            Matcher matcher = pattern.matcher(durationStr);
+
+            if (!matcher.matches()) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT,
+                        "duration is not a valid data duration string: " + durationStr);
+            }
+
+            String valueStr = matcher.group(1);
+            String unitString = matcher.group(2);
+
+            BigDecimal value = new BigDecimal(valueStr);
+
+            // Convert to milliseconds based on the unit
+            BigDecimal milliseconds;
+            switch (unitString) {
+                case "ns":
+                    milliseconds = value.divide(BigDecimal.valueOf(1_000_000), RoundingMode.HALF_UP);
+                    break;
+                case "us":
+                    milliseconds = value.divide(BigDecimal.valueOf(1_000), RoundingMode.HALF_UP);
+                    break;
+                case "ms":
+                    milliseconds = value;
+                    break;
+                case "s":
+                    milliseconds = value.multiply(BigDecimal.valueOf(1_000));
+                    break;
+                case "m":
+                    milliseconds = value.multiply(BigDecimal.valueOf(60_000));
+                    break;
+                case "h":
+                    milliseconds = value.multiply(BigDecimal.valueOf(3_600_000));
+                    break;
+                case "d":
+                    milliseconds = value.multiply(BigDecimal.valueOf(86_400_000));
+                    break;
+                default:
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT,
+                            "Unknown time unit: " + unitString);
+            }
```

Comment:
```suggestion
            BigDecimal value = new BigDecimal(matcher.group(1));
            TimeUnit timeUnit = valueOfTimeUnit(matcher.group(2));

            BigDecimal milliseconds = value.multiply(BigDecimal.valueOf(millisPerTimeUnit(timeUnit)));
            try {
                return milliseconds.longValueExact();
            }
            catch (ArithmeticException e) {
                // If exact conversion fails, use rounding
                return milliseconds.setScale(0, RoundingMode.HALF_UP).longValue();
            }
```

Seems we could always use BigDecimal multiplication here -- since it performs better than division -- through extracting a method similar to `millisPerTimeUnit` in `Duration`.